core/src/main/java/org/apache/iceberg/AllManifestsTable.java

```diff
@@ -204,4 +228,178 @@ public Iterable<FileScanTask> split(long splitSize) {
+
+    private Expression boundExpr;
+    private Long snapshotId;
+
+    private static final boolean ROWS_MIGHT_MATCH = true;
+    private static final boolean ROWS_CANNOT_MATCH = false;
+
+    private SnapshotEvaluator(Expression expr, Snapshot snap, Types.StructType structType, boolean caseSensitive) {
+      this.boundExpr = Binder.bind(structType, expr, caseSensitive);
+      this.snapshotId = snap.snapshotId();
+    }
+
+    private boolean eval() {
+      return ExpressionVisitors.visitEvaluator(boundExpr, this);
+    }
+
+    @Override
+    public Boolean alwaysTrue() {
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public Boolean alwaysFalse() {
+      return ROWS_CANNOT_MATCH;
+    }
+
+    @Override
+    public Boolean not(Boolean result) {
+      return !result;
+    }
+
+    @Override
+    public Boolean and(Boolean leftResult, Boolean rightResult) {
+      return leftResult && rightResult;
+    }
+
+    @Override
+    public Boolean or(Boolean leftResult, Boolean rightResult) {
+      return leftResult || rightResult;
+    }
+
+    @Override
+    public <T> Boolean isNull(BoundReference<T> ref) {
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean notNull(BoundReference<T> ref) {
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean isNaN(BoundReference<T> ref) {
+      if (snapshotIdRef(ref)) {
+        return ROWS_CANNOT_MATCH; // reference_snapshot_id column is never nan
+      } else {
+        return ROWS_MIGHT_MATCH;
+      }
+    }
+
+    @Override
+    public <T> Boolean notNaN(BoundReference<T> ref) {
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean lt(BoundReference<T> ref, Literal<T> lit) {
+      return compare(ref, lit, compareResult -> compareResult < 0);
+    }
+
+    @Override
+    public <T> Boolean ltEq(BoundReference<T> ref, Literal<T> lit) {
+      return compare(ref, lit, compareResult -> compareResult <= 0);
+    }
+
+    @Override
+    public <T> Boolean gt(BoundReference<T> ref, Literal<T> lit) {
+      return compare(ref, lit, compareResult -> compareResult > 0);
+    }
+
+    @Override
+    public <T> Boolean gtEq(BoundReference<T> ref, Literal<T> lit) {
+      return compare(ref, lit, compareResult -> compareResult >= 0);
+    }
+
+    @Override
+    public <T> Boolean eq(BoundReference<T> ref, Literal<T> lit) {
+      return compare(ref, lit, compareResult -> compareResult == 0);
+    }
+
+    @Override
+    public <T> Boolean notEq(BoundReference<T> ref, Literal<T> lit) {
+      return compare(ref, lit, compareResult -> compareResult != 0);
+    }
+
+    @Override
+    public <T> Boolean in(BoundReference<T> ref, Set<T> literalSet) {
+      Comparator<Object> longComparator = Comparators.forType(Types.LongType.get());
+      if (snapshotIdRef(ref)) {
+        boolean in = literalSet.stream().anyMatch(lit -> longComparator.compare(snapshotId, lit) == 0);
```

Comment:
nit: If you want to, you can use `noneMatch` to avoid the negation in `if`:

```
boolean noneMatch = literalSet.stream().noneMatch(lit -> comparator.compare(snapshotId, lit) == 0);
if (noneMatch) {
  return ROWS_CANNOT_MATCH;
}
```